text/0000-windows-security-identifiers.md

```diff
@@ -0,0 +1,257 @@
+
+Also, current platforms/implementations only accept POSIX UID/GID but already convert them into roughly-similar SID to create Windows images. This implicit conversion will inevitably lead to confusion and bugs. Directly using SIDs for Windows images is simpler (and closer to the Linux implementation). Additionally, Windows SIDs are much more intuitive for security-concious Windows users.
+
+> What use cases does it support?
+ 
+This allows platforms/implementations to create Windows images on any OS runtime environment and supports all the same UID/GID ownership use-cases that Linux currently does.
+
+> What is the expected outcome?
+
+From a user perspective, Windows stack authors will specify `CNB_OWNER_SID`/`CNB_GROUP_SID` in stack images and all generated user-owned files in builder, buildpackage, and app images would have the SID value set as their *Owner SID/Group SID*. When the env vars SIDs are not present or invalid for Windows images, platforms/implementations should generate helpful errors.
+
+# What it is
+[what-it-is]: #what-it-is
+
+This provides a high level overview of the feature.
+
+> Define any new terminology.
+
+No new terminology - just existing Windows and Docker security terms.
+
+> Define the target persona: buildpack author, buildpack user, platform operator, platform implementor, and/or project contributor.
+
+Since this involves changing user-facing variables (in stack images) and tightens security in generated images, then stack authors, platform operators, platform implementors, and project contributors would be potentially impacted and would need to be aware of this change.
+
+> Explaining the feature largely in terms of examples.
+  
+As a Windows stack author, so I can control the permissions implementation/platform generated files to minimize the attack vector within my containers, I set the env vars `CNB_OWNER_SID`/`CNB_GROUP_SID` and `USER` to the desired container runtime SID/user. In practice, most stack authors will use Docker's builtin `ContainerUser` (SID: `S-1-5-93-2-1`).
+
+As any other end-user, I use the stack images as usual.
+
+> If applicable, provide sample error messages, deprecation warnings, or migration guidance.
+
+All migration needs should be covered by an API bump and for implementations, changing the values they use for `-uid` and `-gid` to SID strings. Platforms that accept older API versions will need to continue to convert `CNB_USER_ID` into equivalent Security descriptors. See detailed migration steps below ([High-level migration steps](#high-level-migration-steps))
+
+> If applicable, describe the differences between teaching this to existing users and new users.
+
+SID usage will be very similar to Linux, though almost always using the SID for the standard Docker Windows user `ContainerUser`. Any Linux-specific documentation or existing Windows samples for Windows run images would need to change to reflect the new variables for Windows.
```

Comment:
It it worth having `create-builder` automatically set `CNB_USER_SID` and `CNB_GROUP_SID` if it find that the build image is using one of the well known users? This might reduce burden on end users.

Reply:
Great suggestion, particularly since Docker's [hard-coded](https://github.com/moby/moby/blob/751d5f879a4f625bd32b08481bbb45e5d2db5b6c/pkg/system/syscall_windows.go#L50-L51) `ContainerUser`/`ContainerAdministrator` SIDs that will most commonly used, aren't well-known Windows SIDs. Having most stacks follow this pattern could also give us a mitigation if the SIDs ever changed for some.

I'll update the RFC with this suggestion

Reply:
I was going to add some text for that option but was wondering if it starts to contradict valid build image requirements of the spec:

Under [**Build Image**](https://github.com/buildpacks/spec/blob/313078611d8a7925cb69a241df58e7d749d7f364/platform.md#:~:text=CNB_USER_ID%20set%20to%20the%20user)
> The platform MUST ensure that:
>
> The image config's Env field has the environment variable CNB_USER_ID set to the user †UID/‡SID of the user specified in the User field.

(We'll be tweaking the var name here of course)

But I'm now feeling like, while this helps with `pack` builders, this may cause some delayed confusion in a case where a Stack Author makes a "build image" w/o the env vars, that works with `pack` but, potentially much later, a platform like `kpack` fails to validates the original build image while trying to make a cluster builder or something.

Perhaps, instead we could add a more helpful error message during `pack builder create` - when the env vars are missing, suggest the value to use, based on a lookup table for `ContainerUser`, `ContainerAdministrator` (and perhaps even a `/etc/passwd` check for Linux images, if we wanted to go that far).

Reply:
Do we need to close the loop on this?